io.c

```diff
@@ -14700,47 +14700,286 @@ set_LAST_READ_LINE(VALUE val, ID _x, VALUE *_y)
 /*
  * Document-class:  ARGF
  *
- * ARGF is a stream designed for use in scripts that process files given as
- * command-line arguments or passed in via STDIN.
+ * :markup: markdown
  *
- * The arguments passed to your script are stored in the +ARGV+ Array, one
- * argument per element. ARGF assumes that any arguments that aren't
- * filenames have been removed from +ARGV+. For example:
+ * ## \ARGF and `ARGV`
  *
- *     $ ruby argf.rb --verbose file1 file2
+ * The \ARGF object works with the array at global variable `ARGV`:
  *
- *     ARGV  #=> ["--verbose", "file1", "file2"]
- *     option = ARGV.shift #=> "--verbose"
- *     ARGV  #=> ["file1", "file2"]
+ * - **ARGV** may be thought of as the <b>argument vector</b> array.
  *
- * You can now use ARGF to work with a concatenation of each of these named
- * files. For instance, ARGF.read will return the contents of _file1_
- * followed by the contents of _file2_.
+ *     Initially, it contains the arguments and options
+ *     that are passed to it from the command line;
+ *     the running Ruby program can modify that array as it likes.
  *
- * After a file in +ARGV+ has been read ARGF removes it from the Array.
- * Thus, after all files have been read +ARGV+ will be empty.
+ * - **ARGF** may be thought of as the <b>argument files</b> object.
  *
- * You can manipulate +ARGV+ yourself to control what ARGF operates on. If
- * you remove a file from +ARGV+, it is ignored by ARGF; if you add files to
- * +ARGV+, they are treated as if they were named on the command line. For
- * example:
+ *     It can access file streams and/or the `$stdin` stream,
+ *     based on what it finds in `ARGV`.
+ *     This provides a convenient way for the command line
+ *     to specify streams for a Ruby program to read.
  *
- *     ARGV.replace ["file1"]
- *     ARGF.readlines # Returns the contents of file1 as an Array
- *     ARGV           #=> []
- *     ARGV.replace ["file2", "file3"]
- *     ARGF.read      # Returns the contents of file2 and file3
+ * ## Reading
  *
- * If +ARGV+ is empty, ARGF acts as if it contained <tt>"-"</tt> that
- * makes ARGF read from STDIN, i.e. the data piped or typed to your
- * script. For example:
+ * \ARGF may read from _source_ streams,
+ * which at any particular time are determined by the content of `ARGV`.
  *
- *     $ echo "glark" | ruby -e 'p ARGF.read'
- *     "glark\n"
+ * ### Simplest Case
+ *
+ * When the <i>very first</i> \ARGF read occurs with an empty `ARGV` (`[]`),
+ * the source is `$stdin`:
+ *
+ * - \File `t.rb`:
+ *
+ *     ```
+ *     p ['ARGV', ARGV]
+ *     p ['ARGF.read', ARGF.read]
+ *     ```
+ *
+ * - Commands and outputs
+ *   (see below for the content of files `foo.txt` and `bar.txt`):
+ *
+ *     ```sh
+ *     echo "Open the pod bay doors, Hal." | ruby t.rb
```

Comment:
Nit: we should add a leading `$` to differentiate the command and the output.

```suggestion
 *     $ echo "Open the pod bay doors, Hal." | ruby t.rb
```

Reply:
I'd love to do this throughout, but the colorizer does not do even the first line.  Lemme no what you think.

Reply:
I note that in this markdown, the syntax highlighting occurs only for the simplest of command lines (no options or pipe), and in one spot, not even then.  This is very disappointing, b/c my main reason for writing this revision in markdown was to get the highlighting.

I could do one of these:
- Add `$ ` prompts to the example command lines, in which case we might consider switching this document from markdown to rdoc.
- Separate each "Command and Output" section into two sections, "Command" and "Output."

Reply:
I see. Maybe it's not too important to have syntax highlighting for shell commands and we can just switch back to RDoc again? Using RDoc over markdown would also make the documentation style more consistent with the rest of the code base.

Reply:
Prompts added to example shell commands;  format changed from markdown to rdoc.